packages/bull-vault/src/FlashBull.sol

```diff
@@ -141,6 +168,38 @@ contract FlashBull is UniBull {
         IERC20(bullStrategy).transfer(msg.sender, IERC20(bullStrategy).balanceOf(address(this)));
     }
 
+    function flashWithdraw(FlashWithdrawParams calldata _params) external {
+        IERC20(bullStrategy).transferFrom(msg.sender, address(this), _params.bullAmount);
+
+        uint256 usdcToRepay;
+        uint256 crabToRedeem;
+        uint256 wPowerPerpToRedeem;
+
+        {
+            uint256 bullShare = _params.bullAmount.wdiv(IERC20(bullStrategy).totalSupply());
+            crabToRedeem = bullShare.wmul(IERC20(crab).balanceOf(bullStrategy));
+            (, uint256 squeethInCrab) = IBullStrategy(bullStrategy).getCrabVaultDetails();
+            uint256 crabTotalSupply = IERC20(crab).totalSupply();
+            wPowerPerpToRedeem = crabToRedeem.wmul(squeethInCrab).wdiv(crabTotalSupply);
+            usdcToRepay = IBullStrategy(bullStrategy).calcUsdcToRepay(bullShare);
+        }
+
+        // oSQTH-ETH swap
+        _exactOutFlashSwap(
+            weth,
+            wPowerPerp,
+            _params.wPowerPerpPoolFee,
+            wPowerPerpToRedeem,
+            _params.maxEthForSqueeth,
+            uint8(FLASH_SOURCE.FLASH_SWAP_WPOWERPERP),
+            abi.encodePacked(_params.bullAmount, crabToRedeem, wPowerPerpToRedeem, usdcToRepay, _params.maxEthForUsdc, uint256(_params.usdcPoolFee))
```

Comment:
Are these packed in the right order? 

    struct FlashWithdrawParams {
        uint256 bullAmount;
        uint256 maxEthForSqueeth;
        uint256 maxEthForUsdc;
        uint24 wPowerPerpPoolFee;
        uint24 usdcPoolFee;
    }